cedar-policy-core/src/ast/request.rs

```diff
@@ -309,28 +316,28 @@ impl Context {
             .from_json_file(json)
     }
 
-    /// Iterate over the (key, value) pairs in the `Context`; or return `None`
-    /// if the `Context` is purely unknown
+    /// Get an iterator over the (key, value) pairs in the `Context`, cloning
+    /// only if necessary.
     //
-    // PANIC SAFETY: This is safe due to the invariant on `self.context`, `self.context` must always be a record
-    pub fn iter<'s>(&'s self) -> Option<Box<dyn Iterator<Item = (&SmolStr, PartialValue)> + 's>> {
+    // PANIC SAFETY: This is safe due to the invariant (ContextRecord) on `self.context`
+    pub fn values(self) -> Box<dyn Iterator<Item = (SmolStr, PartialValue)>> {
         // PANIC SAFETY invariant on `self.context` ensures that it is a record
         #[allow(clippy::panic)]
-        match self.context.as_ref() {
+        match self.context.into() {
             PartialValue::Value(Value {
                 value: ValueKind::Record(record),
                 ..
-            }) => Some(Box::new(
-                record
-                    .iter()
-                    .map(|(k, v)| (k, PartialValue::Value(v.clone()))),
-            )),
-            PartialValue::Residual(expr) => match expr.expr_kind() {
-                ExprKind::Record(map) => Some(Box::new(
-                    map.iter()
-                        .map(|(k, v)| (k, PartialValue::Residual(v.clone()))),
-                )),
-                ExprKind::Unknown(_) => None,
```

Comment:
How is the `Unknown` case handled now? Kind of worried it might hit the `panic!("internal invariant violation: expected a record, got {kind:?}"),` case

Reply:
Great question! I believe the `Context::from_expr` function is used on all paths for building a context, and this function must return a record value or record residual due to the handing of records in `RestrictedEvaluator::partial_interpret_internal`. I updated the comment about the `ContextRecord` invariant to reflect this, but would appreciate if you would double check my reasoning.

Reply:
Looked at the code closer and I agree that this match is safe given the invariant. 

#1026 introduced `from_partial_value_unchecked` which makes the invariant depend on those call sites. I'd be happier if that didn't need the `_unchecked` constructor, but I am convinced that the invariant is maintained. We already have #541 for future refactoring.

Reply:
Latest version adds additional comments to `from_partial_value_unchecked` to clarify 🙂

Reply:
Would it make sense to implement #541 in this PR if it's convenient?

Reply:
I expect that change to be a bit annoying, so I'd rather see it in a follow up PR. Up to Kesha if wants to do it ofc.

Reply:
Yeah, that will be a somewhat involved change, so I'll leave it for future work.